<a href="https://colab.research.google.com/github/stefkong1982/netology.ru/blob/Master/Statistika_v_Python/Sluchajnye_sobytiya_Sluchajnye_velichiny/DZ_Kondrstev_STPY_48_Sluchajnye_velichiny.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнее задание к лекции «Случайные события. Случайные величины»


Задание
Вопросы по заданию
Преподаватель: Олег Булыгин
Домашнее задание к лекции «Случайные события. Случайные величины»

Смоделировать игру против лотерейного автомата типа “777”. Игрок платит 1 руб., после чего выпадает случайное целое число, равномерно распределенное от 0 до 999. При некоторых значениях числа игрок получает выигрыш.

777: 200 руб.

999: 100 руб.

555: 50 руб.

333: 15 руб.

111: 10 руб.

*77: 5 руб.

**7: 3 руб.

*00: 2 руб.

**0: 1 руб.

Вопросы:

Выгодна ли игра игроку?
Сколько в среднем приобретает или теряет игрок за одну игру?

**Дополнительное задание повышенной сложности.**

Теоретически рассчитать средний выигрыш (проигрыш) и сравнить с результатами моделирования

`*` любая цифра

Вероятности рассчитываются исходя из общего числа вариантов, равного 1000 (числа от 000 до 999).

1. Выпадение числа 777, 999, 555, 333, 111 каждое из этих чисел — это 1 уникальный исход из 1000 возможных. Следовательно, вероятность выпадения каждого из этих чисел равна 1/1000.

2. Да, для комбинации *77 может быть 9 возможных комбинаций. В этом случае, первое число может принимать любое значение от 0 до 9 (10 вариантов), второе число - любое значение от 0 до 9 (10 вариантов), за исключением числа 7 (9 вариантов). Таким образом, общее количество комбинаций для *77 равно 9.
Мы исключаем число 7, потому что в этот момент у нас уже есть две 7 на позициях 2 и 3 (считая с конца). Нам необходимо определить все возможные комбинации для *77, где * означает любое число, кроме 7.
Если бы мы учли 7, тогда мы получили бы комбинацию 777, которая не входит в наш исходный запрос, так как в нем явно указано, что первое число может быть любым, но не 7.
Следовательно, вероятность выпадения чисел равна 9/1000.

3. Выпадение числа **7 — Исключаем двузначные или трехзначные числа, которые оканчиваются на 77.Таких будет 10 чисел в диапазоне от 0 до 100 (77, 177, 277, 377, 477, 577, 677, 777,877, 977). Поэтому для чисел, оканчивающихся на 7, остается 90 возможных вариантов.
Следовательно, вероятность выпадения чисел равна 90/1000.

4. Выпадение числа *00 количество возможных исходов — 10.Любая чифра  от 0 до 9 (10 вариантов).
Следовательно, вероятность выпадения чисел равна 10/1000.

5. Выпадение числа **0.

   Аналогично случаю с числом **7, количество возможных исходов — 90.
   Следовательно, вероятность выпадения чисел равна 90/1000.


### Рассчитаем теоретическую вероятность

In [7]:
cost_of_game = 1         # Стоимость игры

expected_value = 0.001*(200 + 100 + 50 + 15 + 10) + (9/1000)*5 + (90/1000)*3 + (10/1000)*2 + (90/1000)*1

# Считаем средний выигрыш или проигрыш
expected_loss = cost_of_game - expected_value

print(f"Средний выигрыш: {expected_value} руб.")
print(f"Средний проигрыш: {expected_loss} руб.")

Средний выигрыш: 0.7999999999999999 руб.
Средний проигрыш: 0.20000000000000007 руб.


### Рассчитаем вероятность из моделирования

In [8]:
import random

total_games = 1000000    # общее количество игр
cost_of_game = 1         # стоимость игры
total_cost = total_games * cost_of_game  # общая стоимость всех игр

# проверка номера и определение выплаты
def check_number(num):
    if num == 777: return 200
    elif num == 999: return 100
    elif num == 555: return 50
    elif num == 333: return 15
    elif num == 111: return 10
    elif num % 100 == 77: return 5
    elif num % 10 == 7: return 3
    elif num % 100 == 0: return 2
    elif num % 10 == 0: return 1
    else: return 0

total_win = sum(check_number(random.randint(0, 999)) for _ in range(total_games))  # общий выигрыш

avg_win = total_win / total_games  # средний выигрыш на игру
avg_loss = cost_of_game - avg_win  # средний проигрыш на игру

print(f"Средний выигрыш: {avg_win} руб.")
print(f"Средний проигрыш: {avg_loss} руб.")


Средний выигрыш: 0.805827 руб.
Средний проигрыш: 0.19417300000000004 руб.


В этом коде мы запускаем `total_games` игр (`total_games = 1000000`). Для каждой игры мы случайно генерируем число от 0 до 999 и проверяем его на соответствие условиям выплат. Затем мы суммируем все выплаты, чтобы получить общий выигрыш.

Средний выигрыш на игру вычисляется как общий выигрыш, поделенный на общее количество игр. Средний проигрыш на игру вычисляется как стоимость игры минус средний выигрыш на игру.


По результатам моделирования и теоретического расчета, мы получили следующие значения:

- Выигрыш: теоретический - 0.8 руб., модельный - 0.806 руб.
- Проигрыш: теоретический - 0.2 руб., модельный - 0.194 руб.

Сравнивая эти значения, можно сделать вывод, что в среднем игрок теряет около 0.2 рубля за каждую игру против лотерейного автомата типа `777`. Таким образом, игра не является выгодной для игрока, и в долгосрочной перспективе ожидается небольшой проигрыш.